File: Graph/BFS_slide.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
#define MAX 100
int V, E;
bool visited[MAX];
int path[MAX];
vector<int> graph[MAX];

void BFS(int s) {
	for (int i = 0; i < V; i++) {
		visited[i] = false;
		path[i] = -1;
	}
	queue<int> q;
	visited[s] = true;
	q.push(s);
	// to be continued
	while (!q.empty()) {
		int u = q.front();
		q.pop();
		for (int i = 0; i < graph[u].size(); i++) {
			int v = graph[u][i];
			if (!visited[v]) {
				visited[v] = true;
				q.push(v);
				path[v] = u;
			}
		}
	}
}

void printPath(int s, int f) {
	int b[MAX];
	int m = 0;
	if (f == s) {
		cout << s;
		return;
	}
	if (path[f] == -1) {
		cout << "No path" << endl;
		return;
	}
	// to be continued
	while (true) {
		b[m++] = f;
		f = path[f];
		if (f == s) {
			b[m++] = s;
			break;
		}
	}
	for (int i = m - 1; i >= 0; i--) {
		cout << b[i] << " ";
	}
}

void printPathRecursion(int s, int f) {
	if (s == f)
		cout << f << " ";
	else {
		if (path[f] == -1)
			cout << "No path" << endl;
		else {
			printPathRecursion(s, path[f]);
			cout << f << " ";
		}
	}
}

int main() {
	int u, v;
	cin >> V >> E;
	for (int i = 0; i < E; i++) {
		cin >> u >> v;
		graph[u].push_back(v);
		graph[v].push_back(u);
	}
	int s = 0;
	int f = 5;
	BFS(s);
	printPath(s, f);
	return 0;
}

File: Graph/BFS_slide.py
from queue import Queue

MAX = 100
V = None
E = None
visited = [False for i in range(MAX)]
path = [0 for i in range(MAX)]
graph = [[] for i in range(MAX)]

def BFS(s):
    for i in range(V):
        visited[i] = False
        path[i] = -1

    q = Queue()
    visited[s] = True
    q.put(s)
    while not q.empty():
        u = q.get()
        for v in graph[u]:
            if not visited[v]:
                visited[v] = True
                q.put(v)
                path[v] = u


def printPath(s, f):
    b = []
    if f == s:
        print(s)
        return

    if path[f] == -1:
        print("No path")
        return

    while True:
        b.append(f)
        f = path[f]
        if f == s:
            b.append(s)
            break
    for i in range(len(b)-1,-1,-1):
        print(b[i], end=' ')


def printPathRecursion(s, f):
    if s == f:
        print(f, end=' ')
    else:
        if path[f] == -1:
            print("No path")
        else:
            printPathRecursion(s, path[f])
            print(f, end=' ')


if __name__ == '__main__':
    V, E = map(int, input().split())
    for i in range(E):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)
    s = 0
    f = 5
    BFS(s)
    printPath(s, f)
